ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java

```diff
@@ -265,19 +258,29 @@
     }
     List<ReplicaId> backupReplicasToCheck = new ArrayList<>(backupReplicas);
     List<ReplicaId> downReplicasToCheck = new ArrayList<>(downReplicas);
-    if (includeNonOriginatingDcReplicas || this.originatingDcName == null) {
-      backupReplicas.forEach(this::addToEndOfPool);
+
+    // Add replicas that are neither in local dc nor in originating dc.
+    backupReplicas.forEach(this::addToEndOfPool);
+
+    if (routerConfig.routerOperationTrackerIncludeDownReplicas) {
+      // Add those replicas deemed by native failure detector to be down
       downReplicas.forEach(this::addToEndOfPool);
-    } else {
-      // This is for get request only. Take replicasRequired copy of replicas to do the request
-      // Please note replicasRequired is 6 because total number of local and originating replicas is always <= 6.
-      // This may no longer be true with partition classes and flexible replication.
-      // Don't do this if originatingDcName is unknown.
-      while (replicaPool.size() < numOfReplicasRequired && backupReplicas.size() > 0) {
-        addToEndOfPool(backupReplicas.pollFirst());
-      }
-      while (replicaPool.size() < numOfReplicasRequired && downReplicas.size() > 0) {
-        addToEndOfPool(downReplicas.pollFirst());
+      // Add those replicas deemed by Helix to be down (offline). This only applies to GET operation.
+      // Adding this logic to mitigate situation where one or more Zookeeper clusters are suddenly unavailable while
+      // ambry servers are still up.
+      if (routerOperation == RouterOperation.GetBlobOperation
+          || routerOperation == RouterOperation.GetBlobInfoOperation) {
+        Set<ReplicaId> offlineReplicas =
+            new HashSet<>(getEligibleReplicas(partitionId, null, EnumSet.of(ReplicaState.OFFLINE)));
+        Set<ReplicaId> originatingReplicas = offlineReplicas.stream().filter(r -> r.getDataNodeId().getDatacenterName().equals(this.originatingDcName)).collect(
```

Comment:
It may be better to do a loop here. Mutating the offlineReplicas list was a little confusing to me and a loop can save a couple list iterations:
```
Set remoteOfflineReplicas
for (replica in offlineReplicas) {
  if (replica in originating DC) {
    numReplicasInOriginatingDc++;
  }
  if (replica in local DC) {
    addToEndOfPool
  } else {
    remoteOfflineReplicas.add(replica)
  }
}
remoteOfflineReplicas.forEach(addToEndOfPool);
```

Reply:
Sure, this looks more straightforward.